server/src-lib/Hasura/GraphQL/Resolve/InputValue.hs

```diff
@@ -95,8 +95,20 @@ asPGColumnTypeAndValueM v = do
     -- whether the result is 'Nothing' or 'Just', which would change the generated query, so we have
     -- to unconditionally mark the query non-reusable.
     | G.isNullable (_aivType v) -> markNotReusable
-    | otherwise                 -> recordVariableUse variableName columnType
-
+    | otherwise                 -> case _aivDefault v of
+        Nothing -> do
+          recordVariableUse variableName columnType ReusableNoDefault
+        Just dv -> do
+          defValueM <- case dv of
+            AGScalar colTy val -> pure $ WithScalarType colTy val
+            AGEnum _ (AGEReference _ maybeValue) -> do
+              let maybeScalarValue = PGValText . RQL.getEnumValue <$> maybeValue
+              pure $ WithScalarType PGText maybeScalarValue
```

Comment:
We might as well skip the construction of `WithScalarType` if we are not using that information or better yet do you think we can abstract out the logic in L86-90 and reuse it here?

Reply:
Right, some code can be either simplified or unified here.

I have now fixed this by abstracting lines 86-90 as you suggest, but I think there is more work to be done here, as typing information now seems to be passed three times: once in the `PGColumnType`, once in the `WithScalarType`, and once in the inner `PGScalarValue`. This really needs to be reworked. I have made a note elsewhere about the idea of using type families for `WithScalarType` to avoid this duplication of type level information.